Less cloning and dynamic dispatch.

SimonSapin opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/craftytrickster/servo/1)

<!-- Reviewable:end -->

SimonSapin committed:
Less cloning and dynamic dispatch.

SimonSapin commented:
@mbrubeck, could you review this? This is on top of https://github.com/servo/servo/pull/9410

mbrubeck commented:
r+

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

_[components/script/dom/cssstyledeclaration.rs, line 151 [r1]](https://reviewable.io:443/reviews/craftytrickster/servo/1#-KEYHRVo_oC2XTbCmeYn:-KEYHRVo_oC2XTbCmeYo:1758774050) ([raw file](https://github.com/craftytrickster/servo/blob/5c11bef9ed9f4272c4f24bb7e977eed9a98bc57a/components/script/dom/cssstyledeclaration.rs#L151)):_
The `itertools` crate has a `map_fn` iterator adapter for the same purpose. I'm not sure whether you'd rather add a dependency or use this one-off implementation.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/craftytrickster/servo/1#-:-KEYHOfLel-IhXbg9g_Q:5174718)_

<!-- Sent from Reviewable.io -->

SimonSapin commented:
It looks like `MapFn` is a `type` alias to `std::iter::Map`, not a new struct. I’ve tried using an `fn` function with `.map()` but couldn’t get it to be `Clone`, maybe because it had lifetime parameters:

``` rust
/home/simon/projects/servo2/components/script/dom/cssstyledeclaration.rs:151:46: 151:75 error: the trait `core::clone::Clone` is not implemented for the type `fn(&'a core::cell::Ref<'_, style::properties::PropertyDeclaration>) -> &'a style::properties::PropertyDeclaration {dom::cssstyledeclaration::CSSStyleDeclaration.CSSStyleDeclarationMethods::GetPropertyValue::deref}` [E0277]
/home/simon/projects/servo2/components/script/dom/cssstyledeclaration.rs:151             let serialized_value = shorthand.serialize_shorthand_to_string(list.iter().map(deref));
                                                                                                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/simon/projects/servo2/components/script/dom/cssstyledeclaration.rs:151:46: 151:75 help: run `rustc --explain E0277` to see a detailed explanation
/home/simon/projects/servo2/components/script/dom/cssstyledeclaration.rs:151:46: 151:75 help: the following implementations were found:
/home/simon/projects/servo2/components/script/dom/cssstyledeclaration.rs:151:46: 151:75 help:   <fn(A) -> Ret as core::clone::Clone>
/home/simon/projects/servo2/components/script/dom/cssstyledeclaration.rs:151:46: 151:75 help:   <extern "C" fn(A) -> Ret as core::clone::Clone>
/home/simon/projects/servo2/components/script/dom/cssstyledeclaration.rs:151:46: 151:75 help:   <unsafe fn(A) -> Ret as core::clone::Clone>
/home/simon/projects/servo2/components/script/dom/cssstyledeclaration.rs:151:46: 151:75 help:   <unsafe extern "C" fn(A) -> Ret as core::clone::Clone>
```

I’ve also implemented `FnMut` manually. It worked, but in addition to being unstable it was more verbose than this since `FnOnce` had to be also implemented separately.
